### PR TITLE
[la7] Add podcast support

### PR DESCRIPTION
**Pull request by nixxo**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] New feature

---

### Description of your *pull request* and other information

Added support for pages containing podcast both single episode and playlist

### CODE REVIEW

File: yt_dlp/extractor/la7.py [hunk 1]
```diff
@@ -57,3 +63,134 @@ def _real_extract(self, url):
...
+            r'title:\s*([\'"])(?P<title>.+?)\1'),
+            webpage, 'title', group='title')
+
+        description = self._html_search_regex((
+            r'<div class="description">(.+?)</div>',
+            r'<div class="description-mobile">(.+?)</div>',
+            r'<div class="box-txt">([^<]+?)</div>',
+            r'<div class="field-content"><p>(.+?)</p></div>'),
+            webpage, 'description', default=None) or self._html_search_meta(
+                'description', webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py:127]: I think this is much more readable, no? Same for others as well
```suggestion
        description = (
            self._html_search_regex(
                (r'<div class="description">(.+?)</div>',
                 r'<div class="description-mobile">(.+?)</div>',
                 r'<div class="box-txt">([^<]+?)</div>',
                 r'<div class="field-content"><p>(.+?)</p></div>'),
                webpage, 'description', default=None)
            or self._html_search_meta('description', webpage))
```

File: yt_dlp/extractor/la7.py [hunk 2]
```diff
@@ -57,3 +63,134 @@ def _real_extract(self, url):
...
+        date_alt = self._search_regex(
+            r'(\d+[\./]\d+[\./]\d+)', title, 'date_alt', default=None)
+        ppn = self._search_regex(
+            r'ppN:\s*([\'"])(?P<ppn>.+?)\1',
+            webpage, 'ppn', group='ppn', default=None)
+        # if the date is not in the title
+        # and title is the same as the show_title
+        # add the date to the title
+        if date and not date_alt and ppn == title.lower():
+            title += ' del %s' % date
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py:152]: why add date to title?

Comment (nixxo) [reply yt_dlp/extractor/la7.py:152]: becuase, as stated in the title and in the tests, one of the podcast available on the website has the episode title identical to the show title and that results in the single episode title to be indistinguishable from one episode to the other (expecially when downloading a playlist)

eg:
https://www.la7.it/otto-e-mezzo/podcast (ep title same as show title)
https://www.la7.it/dimartedi/podcast (normal)

```
λ py -m yt_dlp --get-title https://www.la7.it/otto-e-mezzo/podcast
Otto e Mezzo Sabato
Otto e Mezzo del 26.03.2021
Otto e Mezzo del 25.03.2021
Otto e Mezzo del 24.03.2021
Otto e Mezzo del 23.03.2021
Otto e Mezzo del 22.03.2021
Otto e Mezzo Sabato
Otto e Mezzo del 19.03.2021
Otto e Mezzo del 18.03.2021
Otto e Mezzo del 17.03.2021
```
```
λ py -m yt_dlp --get-title https://www.la7.it/otto-e-mezzo/podcast
Otto e Mezzo Sabato
Otto e Mezzo
Otto e Mezzo
Otto e Mezzo
Otto e Mezzo
Otto e Mezzo
Otto e Mezzo Sabato
Otto e Mezzo
Otto e Mezzo
Otto e Mezzo
```

instead a "normal" podcast titles are:

```
λ py -m yt_dlp --get-title https://www.la7.it/dimartedi/podcast
L'intervista di Giovanni Floris a Pier Luigi Bersani - Puntata del 23/3/2021
L'intervista di Giovanni Floris a Nicola Magrini (Direttore Aifa) - Puntata del 16/3/2021
Gli effetti del Covid sui giovani - Puntata del 9/3/2020
L'intervista di Giovanni Floris all'astrofisica Ersilia Vaudo - Puntata del 23/2/2021
L'avvocato Emmanuela Bertucci ci spiega come funzionano cashback, flat tax, cartelle esattoriali in scadenza e bonus ristrutturazioni
L'intervista di Giovanni Floris a Gennaro Sangiuliano - Puntata del 9/2/2021
L'intervista di Giovanni Floris a Evelina Christillin - Puntata del 2/2/2021
Giovanni Floris racconta la storia del ghetto di Roma - Puntata del 26/01/2021
L'intervista di Giovanni Floris a Giacomo Mameli - Puntata del 19/01/2021
ASCOLTA LA7 - Valerio Lundini
```

File: yt_dlp/extractor/la7.py [hunk 3]
```diff
@@ -57,3 +63,139 @@ def _real_extract(self, url):
...
+            or self._og_search_title(webpage))
+        ppn = self._search_regex(
+            r'window\.ppN\s*=\s*([\'"])(?P<ppn>.+?)\1',
+            webpage, 'ppn', group='ppn', default=None)
+
+        entries = []
+        for episode in re.finditer(
+                r'<div class="container-podcast-property">([\s\S]+?)(?:</div>\s*){3}',
+                webpage):
+            entries.append(self._extract_infos(episode.group(1), ppn=ppn))
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py:199]:
```suggestion
            entries.append(self._extract_info(episode.group(1), ppn=ppn))
```

File: yt_dlp/extractor/la7.py [hunk 4]
```diff
@@ -57,3 +63,139 @@ def _real_extract(self, url):
...
+            'formats': formats,
+            'thumbnail': thumb,
+            'upload_date': unified_strdate(date),
+        }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        return self._extract_infos(webpage, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py:168]:
```suggestion
        return self._extract_info(webpage, video_id)
```

File: yt_dlp/extractor/la7.py [hunk 5]
```diff
@@ -57,3 +63,139 @@ def _real_extract(self, url):
...
+
+
+class LA7PodcastIE(LA7PodcastEpisodeIE):
+    IE_NAME = 'la7.it:podcast'
+    _VALID_URL = r'(https?://)?(www\.)?la7\.it/(?P<id>[^/]+)/podcast$'''
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py]:
```suggestion
    _VALID_URL = r'(https?://)?(www\.)?la7\.it/(?P<id>[^/]+)/podcast/?(?:$|[#?])'
```

File: yt_dlp/extractor/la7.py [hunk 6]
```diff
@@ -57,3 +63,139 @@ def _real_extract(self, url):
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        return self._extract_infos(webpage, video_id)
+
+
+class LA7PodcastIE(LA7PodcastEpisodeIE):
+    IE_NAME = 'la7.it:podcast'
+    _VALID_URL = r'(https?://)?(www\.)?la7\.it/(?P<id>[^/]+)/podcast/?(?:$|[#?])'''
```
Comment (pukkandan) [inline yt_dlp/extractor/la7.py:173]:
```suggestion
    _VALID_URL = r'(https?://)?(www\.)?la7\.it/(?P<id>[^/]+)/podcast/?(?:$|[#?])'
```

**Comment by nixxo** [comment]
Playlist test
```
λ py test\test_download.py TestDownload.test_LA7Podcast
[debug] Loading archive file None
[la7.it:podcast] propagandalive: Downloading webpage
[download] Downloading playlist: Propaganda Live
[info] Writing playlist metadata as JSON to: test_LA7Podcast_propagandalive.info.json
[la7.it:podcast] playlist Propaganda Live: Collected 10 videos; downloading 10 of them
[download] Downloading video 1 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_371130.info.json
[download] Downloading video 2 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_369898.info.json
[download] Downloading video 3 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_368654.info.json
[download] Downloading video 4 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_367473.info.json
[download] Downloading video 5 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_366373.info.json
[download] Downloading video 6 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_365327.info.json
[download] Downloading video 7 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_363958.info.json
[download] Downloading video 8 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_362617.info.json
[download] Downloading video 9 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_361346.info.json
[download] Downloading video 10 of 10
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7Podcast_360269.info.json
[download] Finished downloading playlist: Propaganda Live
.
----------------------------------------------------------------------
Ran 1 test in 0.678s

OK
```

Single episode test
```
λ py test\test_download.py TestDownload.test_LA7PodcastEpisode
[debug] Loading archive file None
[la7.it:pod:episode] 371497: Downloading webpage
[info] Downloading format(s) mp3
[info] Writing video metadata as JSON to: test_LA7PodcastEpisode_371497.info.json
[debug] Invoking downloader on 'https://limone.iltrovatore.it/audio.mp3?source_r=la7&fn=podcast-voicetown-371497.mp3&mp3l=159&mp3pid=voicetown'
[download] Destination: test_LA7PodcastEpisode_371497.mp3
[download] 100% of 10.00KiB in 00:00
.
----------------------------------------------------------------------
Ran 1 test in 0.925s

OK
```

**Pull request merged**